presto-hive/src/test/java/com/facebook/presto/hive/orc/TestOrcBatchPageSourceFactory.java

```diff
@@ -0,0 +1,103 @@
+import org.apache.hadoop.mapred.FileSplit;
+import org.apache.hadoop.mapred.JobConf;
+import org.joda.time.DateTimeZone;
+import org.testng.annotations.Test;
+
+import java.io.File;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+
+import static com.facebook.presto.hive.HiveFileContext.DEFAULT_HIVE_FILE_CONTEXT;
+import static com.facebook.presto.hive.HiveTestUtils.FUNCTION_AND_TYPE_MANAGER;
+import static com.facebook.presto.hive.HiveTestUtils.HDFS_ENVIRONMENT;
+import static com.facebook.presto.hive.HiveTestUtils.SESSION;
+import static com.facebook.presto.orc.OrcEncoding.ORC;
+import static org.apache.hadoop.hive.serde2.objectinspector.primitive.PrimitiveObjectInspectorFactory.javaStringObjectInspector;
+import static org.testng.Assert.assertTrue;
+import static org.testng.Assert.fail;
+
+public class TestOrcBatchPageSourceFactory
+{
+    @Test
+    public void testCreateOrcPageSourceException() throws Exception
+    {
+        Random random = new Random();
+        List<TestOrcBatchPageSourceMemoryTracking.TestColumn> testColumns = ImmutableList.<TestOrcBatchPageSourceMemoryTracking.TestColumn>builder()
+                .add(new TestOrcBatchPageSourceMemoryTracking.TestColumn("p_empty_string", javaStringObjectInspector, () -> "", true))
+                .add(new TestOrcBatchPageSourceMemoryTracking.TestColumn("p_string", javaStringObjectInspector, () -> Long.toHexString(random.nextLong()), false))
+                .build();
+        File tempFile = File.createTempFile("presto_test_orc_batch_page_source_factory", "orc");
+        tempFile.delete();
+        ImmutableList.Builder<HiveColumnHandle> columnsBuilder = ImmutableList.builder();
+        Configuration configuration = new JobConf(new Configuration(false));
+        FileSplit fileSplit = TestOrcBatchPageSourceMemoryTracking.createTestFile(tempFile.getAbsolutePath(),
+                new OrcOutputFormat(), new OrcSerde(), null, testColumns, 1, 1);
+        HiveFileSplit hiveFileSplit = new HiveFileSplit(
+                fileSplit.getPath().toString(),
+                fileSplit.getStart(),
+                fileSplit.getLength(),
+                fileSplit.getLength(),
+                Instant.now().toEpochMilli(),
+                Optional.empty(),
+                ImmutableMap.of(),
+                0);
+        OrcReaderOptions orcReaderOptions = null; //need to be set to null to trigger PrestoException
+        Optional<EncryptionInformation> encryptionInformation =
+                Optional.of(EncryptionInformation.fromEncryptionMetadata(DwrfEncryptionMetadata.forPerField(
+                ImmutableMap.of("field1", "test1".getBytes()),
+                ImmutableMap.of(),
+                "test_algo",
+                "test_provider")));
+        try (ConnectorPageSource connectorPageSource = OrcBatchPageSourceFactory.createOrcPageSource(ORC, HDFS_ENVIRONMENT,
```

Comment:
The problem with this unit test is that it does not read from the page source it created.

The reason why the page source factory is implemented as it's is that the life-cycle of the OrcDataSource is attached to the life-cycle of the created OrcPageSource. If you use try-catch-with-resource it will close the OrcDataSource right after creating the OrcPageSource and you won't be able to read anything from it.

Reply:
Why does this matter? The idea of the unit test was to make sure a RuntimeException was thrown and it was still caught and wrapped by a PrestoException as expected. The ConnectorPageSource object fails to get instantiated due to the orcReaderOptions.